### PR TITLE
Update NAT Address Discovery Design Document 

### PR DESCRIPTION
**Pull request by jdslab**
Updated design document for NAT support as announced in #4630.

### CODE REVIEW

**Comment by jiceatscion** [comment]
This change is [Reviewable](https://reviewable.io/reviews/scionproto/scion/4659)

**Comment by tzaeschke** [comment]
*[`doc/dev/design/NAT-address-discovery.rst` line 145 at r1](https://reviewable.io/reviews/scionproto/scion/4659#-ODBYPnOE5T58zZdgH5Y:-ODBYPnP3GeupntnD5TY:b-tumags) ([raw file](https://github.com/scionproto/scion/blob/219dd8553bbb86362c46f2bd82062715d88c3a9e/doc/dev/design/NAT-address-discovery.rst#L145)):*
> ```reStructuredText
> --------
> During the open-source contributors meeting on Nov. 19, 2024, it was agreed that the STUN/UDP/IP solution is preferred
> due to its simplicity. However, arguments about message integrity/authentication have not yet been discussed at that time.
> ```

As I remember it, the advantage wasn't simplicity (IP/UDP/SCION/STUN would also be simple) but a general viewpoint that STUN should operate on the underlay because it solves an underlay problem.

**Review by tzaeschke — COMMENTED** [review]
:lgtm:

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4659#-:-ODBaLc25dOhyk3cm6pm:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @jdslab)

**Review by lukedirtwalker — APPROVED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4659)* status: all files reviewed, 1 unresolved discussion (waiting on @jdslab, @jiceatscion, @JordiSubira, and @oncilla)

**Review by lukedirtwalker — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4659#-:-ODPHUxV6s6UbHdTkCqu:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @jdslab, @jiceatscion, @JordiSubira, and @oncilla)

**Review by tzaeschke — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4659)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @jiceatscion, @JordiSubira, and @oncilla)

**Review by tzaeschke — COMMENTED** [review]
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4659)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @jiceatscion, @JordiSubira, and @oncilla)

**Review by marcfrei — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4659)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @jdslab, @jiceatscion, @JordiSubira, @lukedirtwalker, @oncilla, and @tzaeschke)

___
*[`doc/dev/design/NAT-address-discovery.rst` line 69 at r3](https://reviewable.io/reviews/scionproto/scion/4659#-ODuGsGu1MNY-wrv--9B:-ODuGsGv8wuOArxEcd-F:bw1kqbf) ([raw file](https://github.com/scionproto/scion/blob/1476173aeebc400502016b47da840d796feab544/doc/dev/design/NAT-address-discovery.rst#L69)):*
> ```reStructuredText
> 
> Also note that the NAT issue only exists for traffic between separate ASes. If both the client and the server are located within
> the same AS, the server can simply send returning packets to the underlay source address of the client (which is visible to the server in this case).
> ```

Add a note that this is unfortunately only possible for servers that are listening on ports which don't get routed through a SHIM dispatcher.



_Code quote:_
```reStructuredText
Also note that the NAT issue only exists for traffic between separate ASes. If both the client and the server are located within
the same AS, the server can simply send returning packets to the underlay source address of the client (which is visible to the server in this case).
```

**Comment by tzaeschke** [comment]
*[`doc/dev/design/NAT-address-discovery.rst` line 244 at r3](https://reviewable.io/reviews/scionproto/scion/4659#-OE0QXNt52GnrALS8S6y:-OE0QXNt52GnrALS8S6z:b-w2b560) ([raw file](https://github.com/scionproto/scion/blob/1476173aeebc400502016b47da840d796feab544/doc/dev/design/NAT-address-discovery.rst#L244)):*
> ```reStructuredText
> - AS local traffic:
>   For communication within the same AS, the endhost should send returning packets to the underlay source address of the sender
>   simply by reversing src/dst addresses of the underlay in addition to reversing the src/dst addresses in the SCION header.
> ```

I would probably not mention any of the reversing, the reversing is just obvious and hasn't changed in any way. Instead, I would probably emphasize the "underlay":, e.g. "For communication...    to the underlay source address instead of the SCION source address."

**Comment by tzaeschke** [comment]
*[`doc/dev/design/NAT-address-discovery.rst` line 237 at r3](https://reviewable.io/reviews/scionproto/scion/4659#-OE0RuGYE6lOXSKyuujD:-OE0RuGYE6lOXSKyuujE:bjp5ypc) ([raw file](https://github.com/scionproto/scion/blob/1476173aeebc400502016b47da840d796feab544/doc/dev/design/NAT-address-discovery.rst#L237)):*
> ```reStructuredText
> STUN/UDP/IP, STUN/SCION/UDP/IP, and SCMP message extension.
> It was agreed that a PR would be created for the STUN/UDP/IP variant.
> The STUN specification used is the newer specification (RFC5389).
> ```

Insert space: "RFC 5389"'? 
And possibly a link?

**Comment by tzaeschke** [comment]
*[`doc/dev/design/NAT-address-discovery.rst` line 69 at r3](https://reviewable.io/reviews/scionproto/scion/4659#-ODuGsGu1MNY-wrv--9B:-OE0SGJm8V3zOfOnAaQH:b-rqfv8v) ([raw file](https://github.com/scionproto/scion/blob/1476173aeebc400502016b47da840d796feab544/doc/dev/design/NAT-address-discovery.rst#L69)):*
<details><summary><i>Previously, tzaeschke (Tilmann) wrote…</i></summary><blockquote>

By the time this goes into a release, there are hopefully no more dispatchers :-)
</blockquote></details>

I think the NAT issue also exists with intra-AS, it is just solved differently. I would rephrase it to something like: "For intra-AS traffic, the NAT problem can be solved be changing the implementation servers such that return packets are always sent to the underlay address instead (which may be a NATed address)of the SCION source address."

**Review by lukedirtwalker — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4659)* status: all files reviewed, 3 unresolved discussions (waiting on @jdslab, @jiceatscion, @JordiSubira, @marcfrei, and @oncilla)

**Review by jiceatscion — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4659)* status: all files reviewed, 3 unresolved discussions (waiting on @jdslab, @JordiSubira, @marcfrei, and @oncilla)

___
*[`doc/dev/design/NAT-address-discovery.rst` line 69 at r3](https://reviewable.io/reviews/scionproto/scion/4659#-ODuGsGu1MNY-wrv--9B:-OEURDVU8oUl_L33bY_u:bfe0ejj) ([raw file](https://github.com/scionproto/scion/blob/1476173aeebc400502016b47da840d796feab544/doc/dev/design/NAT-address-discovery.rst#L69)):*
<details><summary><i>Previously, tzaeschke (Tilmann) wrote…</i></summary><blockquote>

I think the NAT issue also exists with intra-AS, it is just solved differently. I would rephrase it to something like: "For intra-AS traffic, the NAT problem can be solved be changing the implementation servers such that return packets are always sent to the underlay address instead (which may be a NATed address)of the SCION source address."
</blockquote></details>

I agree that the NAT issue does exist just as well intra-AS: Sunrise has SCION, but I doubt it does me much good with the non-routable address they give me. However, can't we assume that the exact same STUN tactics can be used intra-AS? That is: use the nearest BR as a stun server - on the inside? I know you'll tell me that if the client is behind a symmetric NAT, then it won't work. But that's fine, because if an AS is configured like that internally, then it just violates the SCION protocol entirely: the underlay address is supposed to be a valid endhost address; not some random thing. Am I missing something? I am not too fond of the idea that we can skirt the issue by reversing the underlay header; that assumes that in all cases the underlay header remains associated with the rest of the SCION packet until such time that a response is warranted. It also excludes any kind of NAT-ed  SCION server, which sucks exactly as much as it does in the IP world.

**Review by jiceatscion — COMMENTED** [review]
My objection to the intra-AS solution shouldn't be viewed as blocking. Intra-AS NAT support can be introduced later and need not be tied to the inter-As solution.

> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4659#-:-OEUXjPyB_VzqZbiqlym:b-nc43bc)* status: all files reviewed, 3 unresolved discussions (waiting on @jdslab, @JordiSubira, @marcfrei, and @oncilla)

**Review by jiceatscion — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4659#-:-OEUYBDgERiG2K6ZZUJr:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @jdslab, @JordiSubira, @marcfrei, and @oncilla)

**Comment by tzaeschke** [comment]
> > I think the NAT issue also exists with intra-AS, it is just solved differently. I would rephrase it to something like: "For intra-AS traffic, the NAT problem can be solved be changing the implementation servers such that return packets are always sent to the underlay address instead (which may be a NATed address)of the SCION source address."
> 
> I agree that the NAT issue does exist just as well intra-AS: Sunrise has SCION, but I doubt it does me much good with the non-routable address they give me. However, can't we assume that the exact same STUN tactics can be used intra-AS? That is: use the nearest BR as a stun server - on the inside? I know you'll tell me that if the client is behind a symmetric NAT, then it won't work. But that's fine, because if an AS is configured like that internally, then it just violates the SCION protocol entirely: the underlay address is supposed to be a valid endhost address; not some random thing. Am I missing something? I am not too fond of the idea that we can skirt the issue by reversing the underlay header; that assumes that in all cases the underlay header remains associated with the rest of the SCION packet until such time that a response is warranted. It also excludes any kind of NAT-ed SCION server, which sucks exactly as much as it does in the IP world.

Symmetric NAT may be one issue, but another one is that an AS may be divided into multiple subnets separated by NATs. That means, different border routers may be seeing different NATed addresses. The problem is then that we need to figure out which border router is in the same subnet as the destination host, and there may not be any border router in that subnet. 

The only way I see to avoid using the underlay is having endhosts implement STUN servers, so before I send anything to a destination host, I first STUN them to get my NATed address. Compared to that, it seems simpler to me to use the underlay, but we can discuss this if people disagree?

**Review by jiceatscion — COMMENTED** [review]
Yes the possibility of adding stun server to every endpoint crossed my mind. Sounds brutal, but might not be that bad. It'd be basically the same code, just enabled in more places. Yet, I have a hard time imagining what internal network topology would make that necessary. May be it'd be worth showing an example? As I said, we don't really need to solve it right now. It could be a subsequent effort... even if it leads to generalizing the stun server to every end-point, it would still reuse the stun server code from the inter-AS case, so not much would be wasted.

> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4659#-:-OEYh5066-YUVzsML5f5:b-a6dd0u)* status: all files reviewed, 3 unresolved discussions (waiting on @jdslab, @JordiSubira, @marcfrei, and @oncilla)

**Comment by tzaeschke** [comment]
@jiceatscion 
Okay, so:
1) Dragging along the underlay address: In JPAN, I currently do the following, when I receive a packet, I attach the inverted path to it. If the packet comes from intra-AS, the path is [], so I store the underlay address in the firstHop field. I don't need an extra attribute. When sending an answer, I simply send to the firstHop underlay, no extra logic required (except for ensuring the correct port range).
2) I would actually solve intra-AS and inter-AS together, if possible. It would be weird if one works and the other one doesn't. What do others think?
3) An example for an AS with multiple subnets is the Geant AS, for example Geant Paris and Geant Frankfurt are in the same AS but in different subnets (thanks to @FR4NK-W for providing the example; and yes there isn't really a NAT inbetween them, as I wrote), that may have caused confusion).

Does that make sense?
Should we solve intra/inter-AS separately?

**Review by lukedirtwalker — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4659)* status: all files reviewed, 3 unresolved discussions (waiting on @jdslab, @JordiSubira, @marcfrei, and @oncilla)

**Review by JordiSubira — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4659)* status: all files reviewed, 3 unresolved discussions (waiting on @jdslab, @marcfrei, and @oncilla)

___
*[`doc/dev/design/NAT-address-discovery.rst` line 69 at r3](https://reviewable.io/reviews/scionproto/scion/4659#-ODuGsGu1MNY-wrv--9B:-OFYozxVDWgKmfFUJIB5:bsixlgr) ([raw file](https://github.com/scionproto/scion/blob/1476173aeebc400502016b47da840d796feab544/doc/dev/design/NAT-address-discovery.rst#L69)):*
<details><summary><i>Previously, jiceatscion wrote…</i></summary><blockquote>

I agree that the NAT issue does exist just as well intra-AS: Sunrise has SCION, but I doubt it does me much good with the non-routable address they give me. However, can't we assume that the exact same STUN tactics can be used intra-AS? That is: use the nearest BR as a stun server - on the inside? I know you'll tell me that if the client is behind a symmetric NAT, then it won't work. But that's fine, because if an AS is configured like that internally, then it just violates the SCION protocol entirely: the underlay address is supposed to be a valid endhost address; not some random thing. Am I missing something? I am not too fond of the idea that we can skirt the issue by reversing the underlay header; that assumes that in all cases the underlay header remains associated with the rest of the SCION packet until such time that a response is warranted. It also excludes any kind of NAT-ed  SCION server, which sucks exactly as much as it does in the IP world.
</blockquote></details>

I would also add that on the long term, we cannot assume how packets will be dispatched to end hosts. Mainly, if we ever have kernel support, packets may get encapsulated again to a fix port, so that they can be routed internally by intra-AS middleware. In that sense, ideally, this solution should work also in this case.

**Comment by tzaeschke** [comment]
*[`doc/dev/design/NAT-address-discovery.rst` line 69 at r3](https://reviewable.io/reviews/scionproto/scion/4659#-ODuGsGu1MNY-wrv--9B:-OFuviNJ6qUP_xC40M1a:byjuw0r) ([raw file](https://github.com/scionproto/scion/blob/1476173aeebc400502016b47da840d796feab544/doc/dev/design/NAT-address-discovery.rst#L69)):*
> use the nearest BR as a stun server

The problem really is, which is the "nearest" STUN server? And in which subnet is it?
Copied from my comment just to keep thing in order (I should have put my comment here in the first place):
An example for an AS with multiple subnets is the Geant AS, for example Geant Paris and Geant Frankfurt are in the 
same AS but in different subnets (thanks to [@FR4NK-W](https://github.com/FR4NK-W) for providing the example; and yes there isn't really a NAT inbetween them, as I wrote, that may have caused confusion).

**Review by tzaeschke — COMMENTED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4659)* status: all files reviewed, 3 unresolved discussions (waiting on @FR4NK-W, @jdslab, and @oncilla)

**Comment by tzaeschke** [comment]
@lukedirtwalker @oncilla @jiceatscion @JordiSubira @marcfrei @FR4NK-W 

Apologies for opening this up again but we came up with a new idea that we may want to discuss.

The idea is to let border routers rewrite the source address/port (replace it with the underlay src address/port) on packets on the way out. Previously this idea was dismissed because it would not work with SPAO where the SCION SRC/DST addresses are protected fields. If we were to change SPAO to allow this kind of rewriting of SRC addresses and SRC ports, the NAT solution could became a whole lot simpler.
- We would not need STUN. THis would remove quite some complexity from clients (keep a table of mappings and handle keepalive etc).
- This would do away with any STUN roudtrips required to detect the external NATed address prior to connection to e border router.
- This would allow easy implementation of SCION-aware NATs that could simply rewrite the SRC address/port.

SPAO may anyway need some changes because it is unclear how it works with DRKeys, specifically how clients behind a NAT can get distinct DRKeys instead of sharing one key (unless sharing is okay).

We think it may be worth discussing this in more detail, maybe in the upcoming open-source meeting?

Comments?

**Comment by jiceatscion** [comment]
> Apologies for opening this up again but we came up with a new idea that we may want to discuss.

No need. Better take the time to arrive at the best possible solution.

> The idea is to let border routers rewrite the source address/port (replace it with the underlay src address/port) on packets on the way out.

That sounds both simple and disturbing. In fact, I found it disturbing for a while that the effect of NAT leaks out of the underlay. Trying to think about it from an abstract networking stack perspective, I wonder if the layering violation isn't already implicit in the SCION forwarding code. We just assume that the link-layer address (since that's what the underlay address is from the SCION network perspective) is contained verbatim in the network address. I guess avoiding any address resolution protocol is a worthy goal, but may be we should accept that there needs to be some form of resolution. In the router, the forwarding to a local host is named "ResolveLocalDst()". so, I'm not the first one to think that some non-trivial "resolution" could be in order. For example we could keep a map that we populate from packets received from local sources. The map would not need to store the common case identity mapping. I can take my answer from next Tuesday's discussion.

> We think it may be worth discussing this in more detail, maybe in the upcoming open-source meeting?

Agree. Consider it added to the agenda.

**Comment by shitz** [comment]
From our perspective, rewriting the SCION host address is not desirable. Having a stable address is hugely beneficial for multiple reasons:

1. It allows for source authentication. As you already mentioned, SPAO and DRKey rely on stable source addresses. Introducing a different stable identifier for this purpose introduces all kinds of complexity.
2. At Anapaya, we are currently building a system that enables SCION ASes to authorize endpoints in their AS to use the local SCION infrastructure. Furthermore, this system also does usage accounting to enable monetization of "SCION subscribers". We believe that such an infrastructure is essential for the widespread adoption and deployment of SCION. Commercial ISPs will only deploy SCION if they can monetize it and today this is only possible on the level of SCION links. This system also relies on relatively stable SCION source addresses otherwise it would need to keep much more state for accounting purposes. Specifically we intend that clients resolve the underlay address with STUN and then register themselves at the router (or SCION network access point - SNAP) with a SCION address and the resolved underlay address.
3. NAT has been a source of many issues for all kind of end-to-end properties in the past. If we can avoid poluting the SCION address space with IP NATs I think we should do so even if it comes at the cost of some complexity on the underlay.

For those reasons, we are still advocating for the STUN-based approach.

N.B., once it is ready, we will also create a proposal for this SCION endhost API.
